cpp/ql/lib/semmle/code/cpp/models/implementations/StdContainer.qll

```diff
@@ -55,73 +55,6 @@ private class Vector extends StdSequenceContainer {
   Vector() { this.hasQualifiedName(["std", "bsl"], "vector") }
 }
 
-/**
- * Additional model for standard container constructors that reference the
- * value type of the container (that is, the `T` in `std::vector<T>`).  For
- * example the fill constructor:
- * ```
- * std::vector<std::string> v(100, potentially_tainted_string);
- * ```
- */
-private class StdSequenceContainerConstructor extends Constructor, TaintFunction {
-  StdSequenceContainerConstructor() {
-    this.getDeclaringType() instanceof Vector or
-    this.getDeclaringType() instanceof Deque or
-    this.getDeclaringType() instanceof List or
-    this.getDeclaringType() instanceof ForwardList
-  }
-
-  /**
-   * Gets the index of a parameter to this function that is a reference to the
-   * value type of the container.
-   */
-  int getAValueTypeParameterIndex() {
-    this.getParameter(result).getUnspecifiedType().(ReferenceType).getBaseType() =
-      this.getDeclaringType().getTemplateArgument(0).(Type).getUnspecifiedType() // i.e. the `T` of this `std::vector<T>`
-  }
-
-  /**
-   * Gets the index of a parameter to this function that is an iterator.
-   */
-  int getAnIteratorParameterIndex() { this.getParameter(result).getType() instanceof Iterator }
-
-  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
-    // taint flow from any parameter of the value type to the returned object
-    (
-      input.isParameterDeref(this.getAValueTypeParameterIndex()) or
-      input.isParameter(this.getAnIteratorParameterIndex())
-    ) and
-    (
-      output.isReturnValue() // TODO: this is only needed for AST data flow, which treats constructors as returning the new object
-      or
-      output.isQualifierObject()
-    )
-  }
-}
-
-/**
- * The standard container function `data`.
- */
-private class StdSequenceContainerData extends TaintFunction {
-  StdSequenceContainerData() {
-    this.getClassAndName("data") instanceof Array or
-    this.getClassAndName("data") instanceof Vector
-  }
-
-  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
-    // flow from container itself (qualifier) to return value
-    input.isQualifierObject() and
-    output.isReturnValueDeref()
-    or
-    // reverse flow from returned reference to the qualifier (for writes to
-    // `data`)
-    input.isReturnValueDeref() and
-    output.isQualifierObject()
-  }
-
-  override predicate isPartialWrite(FunctionOutput output) { output.isQualifierObject() }
```

Comment:
Is the information from `isPartialWrite` in the old models encoded in the new models somewhere?

Reply:
That happens automatically due to the way field flow works (the same is true for all CodeQL languages). Consider an example such as:
```cpp
1. std::vector<int> v;
2. v.push_back(source());
3. v.push_back(0);
4. sink(v.at(0));
```
on `main` the [modeledFlowBarrier](https://github.com/github/codeql/blob/40b75342104ec6a449750332af0d81a501cf9ce0/cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll#L802) predicate excludes cases where `isPartialWrite` holds. However, now that this is tracked via field flow, what happens is that:

```mermaid
graph TD;
    A["source()"]-->|storeStep| B["v [post update] on line 2"];
    B-->|SSA| C["v on line 3"];
    C-->|SSA| D["v on line 4"];
    E["0"] -->|storeStep| G["v [post update] on line 3"];
    G -->|SSA| D;
    D -->|readStep| F["v.at(0)"];
```

and this happens completely automatically via the existing interaction between SSA and field flow.